### PR TITLE
feat(Tactic/Linter): add unicode linter for unicode variant-selectors

### PR DESCRIPTION
**Pull request by joneugster**
- add an addition to the unicode linter ensuring variant-selector only appear on specified characters and also ensuring these characters always have the correct variant selector to avoid confusions.
- fix a bug from #34022 which broke autofixing
- some smaller refactors
  - add `String.printCodepointHex` analog to the existing `printCodepointHex` which has been renamed to `Char.printCodepointHex`
  - pass the next char to `findBadUnicodeAux` to make it easier to consider a pair of consecutive characters
  - add some `ToString` instances to the test file to help debugging. (I really needed them badly...)
- automatic fixes flagged by this linter

Unicode characters can be followed by one of the two "variant-selectors" `\FE0E` (text) or `\FE0F` (emoji).
These appear to the user as a single unicode character and they might copy them by accident. For example `✝️` (`\u271d\uFE0F`), `✝` (`\u271d`) and `✝︎` (`\u271d\uFE0E`) are three variants of the "same" unicode character. The one without variant-selector might display as either emoji or not, depending on the user's device and font.
(e.g. for me it is an emoji while editing this message, but a text when previewing)

Everything flagged by this linter can be fixed fully automatically with `lake exe lint-style --fix` or by commenting `bot fix style` on the PR.

---

- tracking PR: #16215

## Testing:

First, drop the last commit from this PR (which includes only the autofixed style issues)

Then call `lake exe lint-style`. You should see 13 new style errors.

Next, add the following line to `nolints-style.txt`:

```
Mathlib/GroupTheory/GroupExtension/Defs.lean : line 26 : ERR_UNICODE_VARIANT : Unexpected unicode variant selector: "↗︎" (U+2197;U+fe0e). Consider deleting it.
```

now `lake exe lint-style` should only display 9 style errors. These can be applied with `lake exe lint-style --fix`.

Next, remove the changes from `nolints-style.txt`. You should then see another 2 errors which can be fixed again.

## Zulip discussions:
- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Whitelist.20for.20Unicode.3F
- https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Lean.20specific.20font

## Related PRs:
- leanprover/lean4#5015 added some emoji-variant selectors to Emojis used in `Lean` itself.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by joneugster** [comment]
bot fix style

File: Mathlib/Tactic/Linter/UnicodeLinter.lean [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+namespace Mathlib.Linter.TextBased.UnicodeLinter
+
+/-- Following any unicode character, this indicates that the emoji-variant should be displayed -/
+def UnicodeVariant.emoji := '\uFE0F'
+
+/-- Following any unicode character, this indicates that the text-variant should be displayed -/
+def UnicodeVariant.text := '\uFE0E'
+
+/-- Prints a unicode character's codepoint in hexadecimal with prefix 'U+'.
+E.g., 'a' is "U+0061".-/
```
Comment (adomani) [inline Mathlib/Tactic/Linter/UnicodeLinter.lean:34]:
```suggestion
E.g., 'a' is "U+0061". -/
```

**Review by YaelDillies — APPROVED** [review]
This looks fine to me. Any meta reviewer to sign off?

maintainer merge

File: Mathlib/Tactic/Linter/UnicodeLinter.lean [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+.ofNat 0x1F4A5,  -- 💥️
+.ofNat 0x1F7E1,  -- 🟡️
+.ofNat 0x1F4A1,  -- 💡️
+.ofNat 0x1F419,  -- 🐙️
+.ofNat 0x1F50D,  -- 🔍️
+.ofNat 0x1F389,  -- 🎉️
+'\u23F3',        -- ⏳️
+.ofNat 0x1F3C1 ] -- 🏁️
+
+/-- Unicode symbols in mathilb that should always be followed by the text-variant selector. -/
```
Comment (adomani) [inline Mathlib/Tactic/Linter/UnicodeLinter.lean:58]:
```suggestion
/-- Unicode symbols in mathlib that should always be followed by the text-variant selector. -/
```

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 3]
```diff
@@ -163,6 +202,21 @@ def parse?_errorContext (line : String) : Option ErrorContext := Id.run do
         | "ERR_SEM" => some (StyleError.semicolon)
         | "ERR_TWS" => some (StyleError.trailingWhitespace)
         | "ERR_WIN" => some (StyleError.windowsLineEnding)
+        | "ERR_UNICODE_VARIANT" => do
+          match (← errorMessage[0]?) with
```
Comment (adomani) [inline Mathlib/Tactic/Linter/TextBased.lean:206]: Relying on word position is the only part that looks a little fragile to me, but this is also already used by the linter anyway.

Could you add a test that specifically flags any change in words of the error message?

**Review by adomani — COMMENTED** [review]
Thanks!

Overall, this looks good to me.

I'll ping @grunweg since he is the main contributor to the text based linters and wants to have a look.

**Comment by kim-em** [comment]
bors delegate grunweg

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 4]
```diff
@@ -66,6 +74,8 @@ inductive ErrorFormat
   | github : ErrorFormat
   deriving BEq
 
+open UnicodeLinter in
+
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:78]:
```suggestion

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 5]
```diff
@@ -74,6 +84,29 @@ def StyleError.errorMessage (err : StyleError) : String := match err with
...
+    else if selector == UnicodeVariant.text then
+      "text"
+    else
+      "default"
+    let oldHex := " ".intercalate <| s.data.map printCodepointHex
+    match s, selector with
+    | ⟨c₀ :: []⟩, some sel =>
+      let newC : String := ⟨[c₀, sel]⟩
+      let newHex := " ".intercalate <| newC.data.map printCodepointHex
+      s!"missing unicode variant-selector at char {pos}: \"{s}\" ({oldHex}). \
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:99]: nit: isn't English grammar to write "variant selector" instead? Can you audit the whole PR for this, please?

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 6]
```diff
@@ -74,6 +84,29 @@ def StyleError.errorMessage (err : StyleError) : String := match err with
...
+      Please use the {variant}-variant: \"{newC}\" ({newHex})!"
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:100]: Similarly, I would say "use the {variant} variant" - also, please check for other instances.

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 3 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/TextBased.lean:206]: Parsing the error message is brittle, but okay enough: please add a test which covers the current wording, though - to make sure to catch a wording change.

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 7]
```diff
@@ -241,12 +295,94 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
...
+    #[(.unicodeVariant ⟨[c]⟩ none 0)] else #[]
+  findBadUnicodeAux s c initalErrors
+
+end UnicodeLinter
+
+/-- Lint a collection of input strings if one of them contains unwanted unicode. -/
+def unicodeLinter : TextbasedLinter := fun lines ↦ Id.run do
+  let mut changed : Array String := #[]
+  let mut errors : Array (StyleError × ℕ) := Array.mkEmpty 0
+  let mut lineNumber := 1
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:352]: The other linters are using this pattern now (where `lineNumber := idx + 1`): I don't have a strong preference between those, but perhaps it's good to be consistent.
```
  for h : idx in [:lines.size] do
    let line := lines[idx]

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 8]
```diff
@@ -241,12 +295,94 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
...
+  for line in lines do
+    let err := UnicodeLinter.findBadUnicode line
+
+    -- try to auto-fix the style error
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:356]:
```suggestion
    -- Try to auto-fix all errors in the current line.

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 9]
```diff
@@ -241,12 +295,94 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
...
+    let mut newLine := line
+    for e in err.reverse do -- reversing is a cheap fix to prevent shifting indices
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:358]: I wonder if this is clearer?
```suggestion
    -- Fix errors from the end, to deal with shifting string indices.
    for e in err.reverse do
```
Another, more verbose option could be
```suggestion
    -- Error fixes may insert or remove variant selectors, hence shift string indices.
    -- The easiest way to deal with this is fixing errors in reverse order.)
    for e in err.reverse do

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 10]
```diff
@@ -241,12 +295,94 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
...
+          -- injecting desired variant-selector
+          head ++ ⟨[s.get 0, v]⟩ ++ tail
+        | none =>
+          -- removing used variant-selector
+          head ++ ⟨[s.get 0]⟩ ++ tail
+      | _ =>
+        unreachable!
+
+    changed := changed.push newLine
+    errors := errors.append (err.map (fun e => (e, lineNumber)))
```
Comment (grunweg) [inline Mathlib/Tactic/Linter/TextBased.lean:374]: use \mapsto?

File: MathlibTest/LintStyleTextBased.lean [hunk 11]
```diff
@@ -0,0 +1,31 @@
+import Mathlib.Tactic.Linter.TextBased
+import Batteries.Data.List.Basic
+
+section unicodeLinter
+
+open Mathlib.Linter.TextBased
+open Mathlib.Linter.TextBased.UnicodeLinter
+
+-- test parsing back error message in `parse?_errorContext` for variant selector errors
```
Comment (grunweg) [inline MathlibTest/LintStyleTextBased.lean:9]: Can you add an error with the text variant also? (And I'm not sure if this test is really flagging all edge cases: for instance the one about the beginning of the line could also be nice to add. Tests can help us, let's use them where applicable.)

**Review by grunweg — COMMENTED** [review]
Thanks for flagging me, @adomani. I gave this a review from the text-based linter perspective, and have mostly just a few nits. Thanks for writing this linter - I can't assess how import the issue it finds is, but this is certainly subtle enough that I did not even know the issue *existed* until this linter.

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 5 above)
Comment (joneugster) [reply Mathlib/Tactic/Linter/TextBased.lean]: done: 7f27e98b3dd558fdfd70cf7e798f746d9335b677

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 6 above)
Comment (joneugster) [reply Mathlib/Tactic/Linter/TextBased.lean:99]: done a95bf2524047562efd0254ea9257639b1084f63a

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 10 above)
Comment (joneugster) [reply Mathlib/Tactic/Linter/TextBased.lean]: Changed it to 

```
errors := errors.append <| err.map ((·, lineNumber))
```

in 7ff43f24afa1dffc4d1411bca2cc5245b33f623a and 6700b73478108dbed368be4464f04597aaf62320

**Comment by leanprover-community-bot-assistant** [comment]
This pull request has conflicts, please merge `master` and resolve them.

File: Mathlib/Tactic/Linter/TextBased.lean [hunk 12]
```diff
@@ -382,7 +473,7 @@ def lintFile (opts : LinterOptions) (path : FilePath) (exceptions : Array ErrorC
         -- check if any exception applies:
         if new_errors.any fun (e, idx) ↦
           (idx - 1 == lineIdx) -- Subtract 1 since linter's line numbers are one-based
-          ∧ (ErrorContext.find?_comparable ⟨e, lineIdx, path⟩ exceptions).isSome
+          ∧ (ErrorContext.find?_comparable ⟨e, lineIdx, path⟩ exceptions).isNone
```
Comment (joneugster) [inline Mathlib/Tactic/Linter/TextBased.lean:476]: This slipped my review of #34022 :(

File: MathlibTest/LintStyleTextBased.lean (hunk 11 above)
Comment (joneugster) [reply MathlibTest/LintStyleTextBased.lean]: done this.

File: Mathlib/Tactic/Linter/TextBased.lean (hunk 3 above)
Comment (joneugster) [reply Mathlib/Tactic/Linter/TextBased.lean]: I've added a comment to the place where the error messages are formulated. Moreover, I've added test the explicitly check the words which are parsed back.

**Comment by joneugster** [comment]
bot fix style

**Comment by adomasbaliuka** [comment]
New PR https://github.com/leanprover-community/mathlib4/pull/34719 because of transitions to PRs from forks.
